// @thi.ng/vectors/random.js
import { normal } from "@thi.ng/random/distributions/normal";
import { SYSTEM } from "@thi.ng/random/system";
import { defHofOp } from "./compile/emit.js";
import { NEW_OUT_A } from "./index.js";
import { normalize } from "./normalize.js";
/**
 * Sets `v` to random vector, with each component in interval `[n..m)`. If no
 * `rnd` instance is given, uses {@link @thi.ng/random#SYSTEM}, i.e.
 * `Math.random`. Creates new vector if `v` is null.
 *
 * @remarks
 * The non-fixed sized version of this function can ONLY be used if `v` is given
 * and initialized to the desired size/length.
 *
 * @param v -
 * @param n - default -1
 * @param m - default 1
 * @param rnd -
 */
export const [random, random2, random3, random4] = defHofOp(SYSTEM, ([a]) => `${a}=rnd.minmax(n,m);`, "a,n=-1,m=1,rnd=op", "a", "a", 0, NEW_OUT_A);
/**
 * Sets `v` to random vector, with each component drawn from given random
 * distribution function (default: gaussian/normal distribution) and scaled to
 * `n` (default: 1). Creates new vector if `v` is null.
 *
 * @remarks
 * The non-fixed sized version of this function can ONLY be used if `v` is given
 * and initialized to the desired size/length.
 *
 * References:
 * - https://docs.thi.ng/umbrella/random/#random-distributions
 * - https://docs.thi.ng/umbrella/random/modules.html#normal
 *
 * @param v -
 * @param rnd -
 * @param n - default 1
 */
export const [randomDistrib, randomDistrib2, randomDistrib3, randomDistrib4] = defHofOp(normal, ([a]) => `${a}=rnd()*n;`, "a,rnd=op(),n=1", "a", "a", 0, NEW_OUT_A);
const $norm = (random) => (v, n = 1, rnd = SYSTEM) => normalize(null, random(v, -1, 1, rnd), n);
const $normDist = (random) => (v, rnd, n = 1) => normalize(null, random(v, rnd), n);
/**
 * Sets `v` to a random vector (using {@link random}), normalized to length `n`
 * (default: 1). If no `rnd` instance is given, uses
 * {@link @thi.ng/random#SYSTEM}, i.e. `Math.random`.
 *
 * @remarks
 * The non-fixed sized version of this function can ONLY be used if `v` is given
 * and initialized to the desired size/length.
 *
 * @param v -
 * @param n -
 * @param rnd -
 */
export const randNorm = $norm(random);
export const randNorm2 = $norm(random2);
export const randNorm3 = $norm(random3);
export const randNorm4 = $norm(random4);
/**
 * Similar to {@link randNorm} but wraps {@link randomDistrib} which draws
 * samples from given distribution function (default: gaussian/normal
 * distribution).
 *
 * @remarks
 * The non-fixed sized version of this function can ONLY be used if `v` is given
 * and initialized to the desired size/length.
 */
export const randNormDistrib = $normDist(randomDistrib);
export const randNormDistrib2 = $normDist(randomDistrib2);
export const randNormDistrib3 = $normDist(randomDistrib3);
export const randNormDistrib4 = $normDist(randomDistrib4);
/**
 * Sets `out` to random vector with each component in the semi-open
 * interval defined by [min,max).
 *
 * @param out -
 * @param min -
 * @param max -
 * @param rnd -
 */
export const [randMinMax, randMinMax2, randMinMax3, randMinMax4] = defHofOp(SYSTEM, ([o, a, b]) => `${o}=rnd.minmax(${a},${b});`, "o,a,b,rnd=op", "o,a,b");
