// @thi.ng/vectors/vec3.js
import { EPS } from "@thi.ng/math/api";
import { declareIndices } from "./compile/accessors.js";
import { MAX3, MIN3, ONE3, X3, Y3, Z3, ZERO3, } from "./api.js";
import { AVec } from "./avec.js";
import { intoStridedBuffer, mapStridedBuffer } from "./buffer.js";
import { eqDelta3 } from "./eqdelta.js";
import { hash } from "./hash.js";
import { stridedValues, vecIterator } from "./iterator.js";
import { setS3 } from "./sets.js";
export class Vec3 extends AVec {
    constructor(buf, offset = 0, stride = 1) {
        super(buf || [0, 0, 0], offset, stride);
    }
    /**
     * Returns array of memory mapped {@link Vec3} instances using given
     * backing array and stride settings: The `cstride` is the step size
     * between individual XYZ vector components. `estride` is the step
     * size between successive vectors. This arrangement allows for
     * different storage approaches, incl. SOA, AOS, striped /
     * interleaved etc.
     *
     * @param buf - backing array
     * @param num - num vectors
     * @param start -  start index
     * @param cstride - component stride
     * @param estride - element stride
     */
    static mapBuffer(buf, num = (buf.length / 3) | 0, start = 0, cstride = 1, estride = 3) {
        return mapStridedBuffer(Vec3, buf, num, start, cstride, estride);
    }
    /**
     * Merges given `src` iterable of {@link Vec3}s into single array `buf`.
     * Vectors will be arranged according to given component and element
     * strides, starting at `start` index. It's the user's
     * responsibility to ensure the target buffer has sufficient
     * capacity to hold the input vectors. See `Vec3.mapBuffer` for the
     * inverse operation. Returns `buf`.
     *
     * @param buf -
     * @param src -
     * @param start -
     * @param cstride -
     * @param estride -
     */
    static intoBuffer(buf, src, start = 0, cstride = 1, estride = 3) {
        return intoStridedBuffer(setS3, buf, src, start, cstride, estride);
    }
    static iterator(buf, num, start = 0, cstride = 1, estride = 3) {
        return vecIterator(Vec3, buf, num, start, cstride, estride);
    }
    [Symbol.iterator]() {
        return stridedValues(this.buf, 3, this.offset, this.stride);
    }
    get length() {
        return 3;
    }
    copy() {
        return new Vec3([this.x, this.y, this.z]);
    }
    copyView() {
        return new Vec3(this.buf, this.offset, this.stride);
    }
    empty() {
        return new Vec3();
    }
    eqDelta(v, eps = EPS) {
        return eqDelta3(this, v, eps);
    }
    hash() {
        return hash(this);
    }
    toJSON() {
        return [this.x, this.y, this.z];
    }
}
Vec3.X_AXIS = new Vec3(X3);
Vec3.Y_AXIS = new Vec3(Y3);
Vec3.Z_AXIS = new Vec3(Z3);
Vec3.MIN = new Vec3(MIN3);
Vec3.MAX = new Vec3(MAX3);
Vec3.ZERO = new Vec3(ZERO3);
Vec3.ONE = new Vec3(ONE3);
declareIndices(Vec3.prototype, ["x", "y", "z"]);
export const vec3 = (x = 0, y = 0, z = 0) => new Vec3([x, y, z]);
export const vec3n = (n) => new Vec3([n, n, n]);
export const asVec3 = (x) => x instanceof Vec3
    ? x
    : new Vec3(x.length >= 3
        ? x
        : [x[0] || 0, x[1] || 0, x[2] || 0]);
